staker/txbuilder/builder.go

```diff
@@ -68,59 +108,45 @@ func (b *Builder) ClearTransactions() {
 	b.transactions = nil
 }
 
-func (b *Builder) EstimateGas(ctx context.Context, call ethereum.CallMsg) (gas uint64, err error) {
-	if len(b.transactions) == 0 && !b.isAuthFake {
-		return b.Client.EstimateGas(ctx, call)
+func (b *Builder) tryToFillAuthAddress() {
+	if b.multiTxAuth.From == (common.Address{}) {
+		if addr := b.wallet.Address(); addr != nil {
+			b.multiTxAuth.From = *addr
+			b.singleTxAuth.From = *addr
+		}
 	}
-	return 0, nil
 }
 
-func (b *Builder) SendTransaction(ctx context.Context, tx *types.Transaction) error {
-	b.transactions = append(b.transactions, tx)
-	err := b.wallet.TestTransactions(ctx, b.transactions)
-	if err != nil {
-		// Remove the bad tx
-		b.transactions = b.transactions[:len(b.transactions)-1]
-		return err
-	}
-	return nil
-}
-
-// While this is not currently required, it's recommended not to reuse the returned auth for multiple transactions,
-// as for an EOA this has the nonce in it. However, the EOA wwallet currently will only publish the first created tx,
-// which is why that doesn't really matter.
-func (b *Builder) AuthWithAmount(ctx context.Context, amount *big.Int) (*bind.TransactOpts, error) {
-	nonce, err := b.NonceAt(ctx, b.builderAuth.From, nil)
-	if err != nil {
-		return nil, err
-	}
-	return &bind.TransactOpts{
-		From:     b.builderAuth.From,
-		Nonce:    new(big.Int).SetUint64(nonce),
-		Signer:   b.builderAuth.Signer,
-		Value:    amount,
-		GasPrice: b.builderAuth.GasPrice,
-		GasLimit: b.builderAuth.GasLimit,
-		Context:  ctx,
-	}, nil
+func (b *Builder) AuthWithAmount(ctx context.Context, amount *big.Int) *bind.TransactOpts {
+	b.authMutex.Lock()
+	defer b.authMutex.Unlock()
+	b.tryToFillAuthAddress()
+	auth := b.multiTxAuth
+	auth.Context = ctx
+	auth.Value = amount
+	return &auth
 }
 
 // Auth is the same as AuthWithAmount with a 0 amount specified.
-// See AuthWithAmount docs for important details.
-func (b *Builder) Auth(ctx context.Context) (*bind.TransactOpts, error) {
+func (b *Builder) Auth(ctx context.Context) *bind.TransactOpts {
 	return b.AuthWithAmount(ctx, common.Big0)
 }
 
-func (b *Builder) Transactions() []*types.Transaction {
-	return b.transactions
-}
-
-// Auth is the same as AuthWithAmount with a 0 amount specified.
-// See AuthWithAmount docs for important details.
-func (b *Builder) BuilderAuth() *bind.TransactOpts {
-	return b.builderAuth
+// SingleTxAuth should be used if you need an auth without the transaction batching of the builder.
+func (b *Builder) SingleTxAuth() *bind.TransactOpts {
```

Comment:
Why does the multiTxAuth returned from the `Auth` and `AuthWithAmount` methods bind the authorization to a context before returning it, but the `SingleTxAuth` doesn't?

Reply:
That's mainly just due to its callsite. The other auth methods are called very close to where the auth is needed, and they can therefore pass in the context they want to use. SingleTxAuth however is passed to bold as-is, and bold then hopefully attaches contexts before the tx opts are used (though I haven't confirmed that). We can't attach a context to the auth ourselves because we don't have a great context to scope it to.